llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,298 @@
+  // TODO: 128 is not a standard bit width, will handle later as it's more complex than simply promoting
+  return BitWidth == 1 || BitWidth == 8 || BitWidth == 16 || BitWidth == 32 || BitWidth == 64 || BitWidth == 128;
+}
+
+unsigned HipPromoteIntsPass::getPromotedBitWidth(unsigned Original) {
+  if (Original <= 8) return 8;
+  if (Original <= 16) return 16;
+  if (Original <= 32) return 32;
+  return 64;
+}
+
+Type* HipPromoteIntsPass::getPromotedType(Type* TypeToPromote) {
+    if (auto* IntTy = dyn_cast<IntegerType>(TypeToPromote)) {
+        unsigned PromotedWidth = getPromotedBitWidth(IntTy->getBitWidth());
+        return Type::getIntNTy(TypeToPromote->getContext(), PromotedWidth);
+    }
+    return TypeToPromote;  // Return original type if not an integer
+}
+
+struct Replacement {
+    Instruction* Old;
+    Value* New;
+    Replacement(Instruction* O, Value* N) : Old(O), New(N) {}
+};
+
+void processInstruction(Instruction *I, Type *NonStdType, Type *PromotedTy, std::string Indent,
+                       SmallVectorImpl<Replacement> &Replacements,
+                       SmallDenseMap<Value*, Value*> &PromotedValues) {
+    IRBuilder<> Builder(I);
+    
+    /// Helper to get or create promoted version of a value
+    auto getPromotedValue = [&](Value* V) -> Value* {
+        // First check if we already promoted this value
+        if (PromotedValues.count(V))
+            return PromotedValues[V];
+        
+        // If it's already the right type, return it
+        if (V->getType() == PromotedTy)
+            return V;
+            
+        // If it's the non-standard type, promote it
+        if (V->getType() == NonStdType) {
+            auto NewV = Builder.CreateZExt(V, PromotedTy);
+            PromotedValues[V] = NewV;
+            return NewV;
+        }
+        
+        // Otherwise return original value
+        return V;
+    };
+
+    if (isa<PHINode>(I)) {
+        PHINode* Phi = cast<PHINode>(I);
+        // Create new PHI node with the promoted type (e.g., i64) instead of original type
+        Type* PromotedType = HipPromoteIntsPass::getPromotedType(Phi->getType());
+        PHINode* NewPhi = PHINode::Create(PromotedType, Phi->getNumIncomingValues(), "", Phi);
+        
+        // Copy all incoming values and blocks
+        for (unsigned i = 0; i < Phi->getNumIncomingValues(); ++i) {
+            Value* IncomingValue = Phi->getIncomingValue(i);
+            BasicBlock* IncomingBlock = Phi->getIncomingBlock(i);
+            
+            // If the incoming value is from our promotion chain, use the promoted value
+            Value* NewIncomingValue = PromotedValues.count(IncomingValue) ? 
+                                    PromotedValues[IncomingValue] : IncomingValue;
+            
+            // If the incoming value isn't promoted yet, promote it now
+            if (NewIncomingValue->getType() != PromotedType) {
+                NewIncomingValue = Builder.CreateZExt(NewIncomingValue, PromotedType);
```

Comment:
I think this potentially inserts the new instruction before another PHI node which is ilegal.

Reply:
Same issue as above, I believe

Reply:
Maybe. If it is a placeholder value that will get deleted later in this pass, then its fine. If it sticks, you may end up with something like:
```
%x = phi ...
%y = zext ...
%z = phi ...
```
which is ill-formed. All PHI instructions must be come first before other instructions.